### ISSUE TITLE
Segfault and corrupt wallet files caused by exporting and importing outputs with monero-wallet-rpc

### ISSUE THREAD

**Issue by woodser** [monero-project/monero#8484]
I'm able to segfault monero-wallet-rpc and corrupt the wallet files by calling export_outputs then import_outputs with the result. The calls succeed, indicating 0 outputs were imported. But then these errors start appearing when the next transaction is sent. My tests send funds back to the original wallet so it might be something specific to that. Otherwise, my build and tests are working on v0.18.0.0.

```
2022-08-04 23:38:41.438	W Spent money: 0.000000000000, with tx: <48219a8a2c1534c13b2d79448e79598192829327686f45fdd2fcc04649976c64>
2022-08-04 23:38:41.438	E Invalid index
2022-08-04 23:38:41.438	E Error parsing blocks: Invalid index
2022-08-04 23:38:46.206	E Public key 2788e1c6b262a767ee5bcaeea8da6a560c8e68c8d5f8c46aaf97f8b54e9fd5a2 from received 0.000000600000 output already exists with unspent 0.000000600000 in tx <48219a8a2c1534c13b2d79448e79598192829327686f45fdd2fcc04649976c64>, received output ignored
2022-08-04 23:38:46.208	E Public key c185f909de3c0ef5db4255ed423060c00ccc09ad379f08bc669dfdd820ee0fdd from received 0.000000600000 output already exists with unspent 0.000000600000 in tx <48219a8a2c1534c13b2d79448e79598192829327686f45fdd2fcc04649976c64>, received output ignored
```

I confirmed the issue is introduced with #8179 and resolved by reverting this PR.

**Comment by moneromooo-monero** [comment]
I could make two txes in a row with cold signing, exporting/importing outputs/key images both times. Can you describe your steps in more detail ?

In any case, reading the code, I see something is missing, below. My test wasn't run with that patch. Untested yet.

```
diff --git a/src/wallet/wallet2.cpp b/src/wallet/wallet2.cpp
index 195763949..d87aa0e33 100644
--- a/src/wallet/wallet2.cpp
+++ b/src/wallet/wallet2.cpp
@@ -13241,7 +13241,7 @@ size_t wallet2::import_outputs(const std::pair<uint64_t, std::vector<tools::wall
     exported_transfer_details etd = outputs.second[i];
     transfer_details &td = m_transfers[i + offset];
 
-    // setup td with "cheao" loaded data
+    // setup td with "cheap" loaded data
     td.m_block_height = 0;
     td.m_txid = crypto::null_hash;
     td.m_global_output_index = etd.m_global_output_index;
@@ -13254,6 +13254,8 @@ size_t wallet2::import_outputs(const std::pair<uint64_t, std::vector<tools::wall
     td.m_key_image_known = etd.m_flags.m_key_image_known;
     td.m_key_image_request = etd.m_flags.m_key_image_request;
     td.m_key_image_partial = false;
+    td.m_subaddr_index.major = etd.m_subaddr_index_major;
+    td.m_subaddr_index.minor = etd.m_subaddr_index_minor;
 
     // skip those we've already imported, or which have different data
     if (i + offset < original_size)
diff --git a/src/wallet/wallet2.h b/src/wallet/wallet2.h
index 16e898ad8..6598dea10 100644
--- a/src/wallet/wallet2.h
+++ b/src/wallet/wallet2.h
@@ -401,9 +401,11 @@ private:
       } m_flags;
       uint64_t m_amount;
       std::vector<crypto::public_key> m_additional_tx_keys;
+      uint32_t m_subaddr_index_major;
+      uint32_t m_subaddr_index_minor;
 
       BEGIN_SERIALIZE_OBJECT()
-        VERSION_FIELD(0)
+        VERSION_FIELD(1)
         FIELD(m_pubkey)
         VARINT_FIELD(m_internal_output_index)
         VARINT_FIELD(m_global_output_index)
@@ -411,6 +413,8 @@ private:
         FIELD(m_flags.flags)
         VARINT_FIELD(m_amount)
         FIELD(m_additional_tx_keys)
+        VARINT_FIELD(m_subaddr_index_major)
+        VARINT_FIELD(m_subaddr_index_minor)
       END_SERIALIZE()
     };
 
```

**Comment by woodser** [comment]
Yeah it's exporting and importing back to the same wallet.

Pretty sure a tx is broadcast which sends funds back to the same account, then wallet rpc `export_outputs` and `import_outputs` are called on the same wallet, then another tx is broadcast sending funds back to the same account.

**Comment by moneromooo-monero** [comment]
The cold wallet doesn't call the daemon, so I don't think the mask is needed by it.
Maybe the best is to error out when importing outputs in a hot wallet. And presumably importing key images in a cold wallet.

**Comment by moneromooo-monero** [comment]
The three patches at https://github.com/moneromooo-monero/bitmonero/tree/io allow me to export outputs, import them in the same wallet, and make a tx/sweep_all to itself. If you still get it to break, please be more precise about the conditions.

**Comment by woodser** [comment]
It seems to be working. I can't recreate the issues after applying the patches. Thanks.

**Comment by moneromooo-monero** [comment]
Thanks. I'll wait to see what j-berman says about mask and PR if it still seems good.

**Comment by j-berman** [comment]
>  Maybe the best is to error out when importing outputs in a hot wallet. And presumably importing key images in a cold wallet.

I agree this seems like the best solution to me too. Left some comments on the commits

**Comment by moneromooo-monero** [comment]
I cannot find those. Mind posting a link ?

**Comment by selsta** [comment]
https://github.com/moneromooo-monero/bitmonero/commits/io

Click on the patches and the comments should show up.

**Comment by moneromooo-monero** [comment]
Thanks. I thought I'd done that and saw just wooder's comment, guess I did it wrong.

**Comment by woodser** [comment]
Just updating that this issue is still present in v0.18.1.0, and if I apply the patches on top of v0.18.1.0, things are stable, except it occasionally hangs on import_outputs (when used correctly with view-only and offline wallets) for some unknown reason, but I didn't investigate where exactly it's hanging.

**Comment by j-berman** [comment]
@woodser can you see if you can repro the hanging with #8510 :)

**Comment by woodser** [comment]
The hanging is resolved with #8513.

**Comment by selsta** [comment]
Issue appears to be resolved.
